petrussola/spacex
# 20210418 task2 favorites

## Overview 

Reference story #ABC

- Add favorite feature: User can mark a launch or a pad as favorite, and these are stored in browser so User can recover list of favorites when they access the app again in the future.
- Favorites are available in a sliding component available from the page with the launches or pads tiles. As per wireframe provided, sliding components slides from the right and User can remove there items from list of favorites.

## Update Part 1 - Typology of Chakra UI components used

- Extensive research of Chakra 0.80 version was done. While a newer version of the library exists (1.5.2), the project is using the 0.8 version and there is no compatibility between both version as the newer version relies on a different set of sister libraries.
- Chakra components Used are:

1. IconButton - Chakra is very explicit in recommending to use IconButton instead of Icon for those use cases where an onclick should be attached. As explained later, an onclick is indeed attached to the icons to allow changing from a 'filled star' (to indicate an item has been favorited) to outline star (default icon while item is not favorites).
2. Drawer - drawer provides the sliding effect and functionality needed for the feature without much customisation. It has quite a lot of properties that enhance the experience. While Chakra docs were not very explicit about the full amount of customization that was available to drawer, it is very similar to how other components like Box or Modal work, and the props that were indicated for these two would work for Drawer. A bit of guess work and trial and error had to be done but eventually the effort pay off, as a way to make the drawer scrollable was found.
3. Chakra works seamlessly with a vast library of icons called React-Icons. Given the relatively small amount of icons, and the fact that the one needed were not available natively in Chakra (no outline of star), decision was taken to introduce this library.

## Update Part 2 - state management

- I decided to introduce new pieces of state, both app-wide and local at the component level, were introduced to allow for information about whether a component was favorited or not to circulate across components and synch them in real time no matter where the action to mark them as favorite took place (item tile or item page).
- Local Storage Browser API was used to store the favorites in the browser. This data would be pulled upon app mounting so it is available across the app from the start.
- A key element of the architecture of this feature is the use of useEffect hook, in order to achieve 1) loding of data upon first mount and 2) sync state across components when action was taken by User (favorite an item) anywhere in the app.
- A shared drawer component is used fro launched and pads. Decision was taken to pass props with names that could be used for either list of favorites and then some conditional logic has been included to display either tiles for launches or pads. I believe this will scale as potentially more and more pages are added to the app (currently only launches and launch pads).

## Update Part 3 - how data and logic flows across components

1. App.js - app loads and sets favorite launches and pads based on the info stored in local storage. Should the list of favorite items in local state change, a useEffect updating local storage would run to ensure local storage always has the most up to date state. Favorite list state lives in App.js and trickles down components using prop drilling. Context API would be adequate should the App scale considerably but decision to keep prop drilling was kept for this MVP of the feature.
2. Launches.js / launch-pads.js - custom hook provided by Chakra to manage opening and closing of the drawer was introduced with useDisclosure https://v0.chakra-ui.com/usedisclosure . 
3.  LaunchItem / LaunchPadItem components. Local state to keep track whether an item was favorited or not. Upon a User clicking on the star icon, the function handler would set state, add the element to the favorite list and display the filled star component or the outline depending on the choice. The fact that favorite list state was changed triggers the App.js useEffect level that saves list of favorites into local storage.
4. Drawer. It used the same components in 3. However, instead of displaying a star, as per design requirements it displays an X to remove item from favorite list. This is achieved by knowing in which real state the item component is being rendered and showing X or star depending whether component is displayed as tile in launches.js / launch-pads.js page or within the drawer.
5. Item Page - Similar logic as in the item component. Local state to keep track of whether item is favorited or not is introduced and the same handler when User click on the star to update state across the board (and local storage) and display one Icon or Another.

## How to test

Head to /launches, click on the star outline, the star will fill, click on Favorites button added on top right corner, teh favorite item will be displayed there, click on X to remove item, the star outline in the tile will be back. Click on the tile to go to the page of the item, click on the star, go back to launches or launch-pads page, the item should have a filled star, click on Favorites again, the item will be there. Close tab, reopen and same favorited items will be there. Same testing can be done for Launches and Launch-Pads.
